pkg/operator/ceph/cluster/cluster.go

```diff
@@ -290,48 +296,52 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
 	return false, ""
 }
 
-func (c *cluster) unsetOrchestrationRunning() {
-	c.orchRunMux.Lock()
-	c.orchestrationRunning = false
-	c.orchRunMux.Unlock()
+func (c *cluster) setOrchestrationNeeded() {
+	c.orchMux.Lock()
+	c.orchestrationNeeded = true
+	c.orchMux.Unlock()
 }
 
-func (c *cluster) setOrchestrationPending() {
-	c.orchPenMux.Lock()
-	c.orchestrationPending = true
-	c.orchPenMux.Unlock()
-}
-
-func (c *cluster) getOrchestrationRunning() bool {
-	c.orchRunMux.Lock()
-	defer c.orchRunMux.Unlock()
-	return c.orchestrationRunning
-}
-
-func (c *cluster) getOrchestrationPending() bool {
-	c.orchPenMux.Lock()
-	defer c.orchPenMux.Unlock()
-	return c.orchestrationPending
+// unsetOrchestrationStatus resets the orchestration flags to its original status when
+// there is no additional orchestration request pending
+func (c *cluster) unsetOrchestrationStatus(err error) {
+	c.orchMux.Lock()
+	defer c.orchMux.Unlock()
+	if c.orchestrationNeeded == false {
+		logger.Infof("There is no new instance specification in namespace %s. Stopping orchestration.", c.Namespace)
+		c.orchestrationRunning = false
+	} else if err != nil {
+		// we need to print an existing error of failed orchestration runs if there is another run pending.
+		// otherwise the error-message would be lost as it gets reset by the next orchestration run.
+		logger.Errorf(err.Error() + ". There is another orchestration pending. Proceeding with next orchestration run")
+	}
 }
 
-// make checking and setting orchestrationRunning "atomic"
-func (c *cluster) checkSetOrchestrationRunning() bool {
-	defer c.orchRunMux.Unlock()
-	c.orchRunMux.Lock()
-	if c.orchestrationRunning == false {
+// checkSetOrchestrationStatus is responsible to do orchestration as long as there is a request needed
+// the thread local variable isRunningThread indicates if the calling function is the initial one which started the
+// orchestration and is running the for-loop.
+func (c *cluster) checkSetOrchestrationStatus(isRunningThread bool) bool {
+	c.orchMux.Lock()
+	defer c.orchMux.Unlock()
+	// check if there is an orchestration needed currently
+	if c.orchestrationNeeded == true && c.orchestrationRunning == false {
+		// there is an orchestration needed and no orchestration running currently
+		// allow to enter the orchestration-loop
+		c.orchestrationNeeded = false
 		c.orchestrationRunning = true
+		return true
+	} else if c.orchestrationNeeded == true && c.orchestrationRunning == true && isRunningThread == true {
+		// there is an orchestration needed and the calling function is the thread which executes the
+		// orchestration-loop
+		// allow to run an additional iteration of the orchestration-loop
+		c.orchestrationNeeded = false
+		return true
+	} else if c.orchestrationNeeded == false && c.orchestrationRunning == true && isRunningThread == true {
```

Comment:
This condition allows a second thread to have taken over orchestration work. If we are allowing another thread to take over, I think the logic in this function and in the `unset...` function are unnecessarily complicated. This can be alleviated by returning from inside the `for` loop instead of handling additional logic for when we are allowed to return after the for loop.

The below should allow another thread to take over, still reports errors to the log, and eliminates the need to follow whether this is the running thread.
```go
func orchestrate() {
    var err error
    for checkSet() {
        if err != nil {
            // there was an error encountered last loop, report it so it's not hidden
            logger.Errorf(...)
        }

        err = doStuff()
        unset()
    }

    return err
}

func checkSet() {
    m.Lock()
    defer m.Unlock()
    if needed == true && running == false {
        return true
    }
    return false
}

func unset() {
    m.Lock()
    defer m.Unlock()
    running = false
}
```

If it is not desirable for another thread to be able to take over, then the `canReturn()` method I mentioned last round would solve the issue I think.

Reply:
You are correct. This is with my last changes not necessary anymore.